src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  const currentState = { ...state };
+
+  for (const action of actions) {
+    const { extraData, type, keysToRemove } = action;
+
+    switch (type) {
+      case 'clear':
+        // eslint-disable-next-line no-const-assign
+        for (const key in currentState) {
+          delete currentState[key];
+        }
+        break;
+
+      case 'addProperties':
+        Object.assign(currentState, extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of keysToRemove) {
+          delete currentState[key];
+        }
+        break;
+
+      default:
+        throw new Error('Error');
+    }
+
+    states.push({ ...currentState });
```

Comment:
The current implementation of pushing the state clone into the 'states' array at the end of each loop iteration is correct, but the name 'currentState' is misleading as mentioned before. After renaming 'currentState', ensure that the clone creation is done correctly without mutations.